README.md
[![GitHub issues](https://img.shields.io/github/issues/Legytma/schema_widget)](https://github.com/Legytma/schema_widget/issues "GitHub issues")
[![GitHub forks](https://img.shields.io/github/forks/Legytma/schema_widget)](https://github.com/Legytma/schema_widget/network "GitHub forks")
[![GitHub stars](https://img.shields.io/github/stars/Legytma/schema_widget)](https://github.com/Legytma/schema_widget/stargazers "GitHub stars")
[![GitHub license](https://img.shields.io/github/license/Legytma/schema_widget)](https://github.com/Legytma/schema_widget/blob/master/LICENSE "GitHub license")
[![Pub](https://img.shields.io/pub/v/schema_widget)](https://pub.dev/packages/schema_widget "Pub")
[![Likes](https://badges.bar/schema_widget/likes)](https://pub.dev/packages/schema_widget/score "Likes")
[![Pub Points](https://badges.bar/schema_widget/pub%20points)](https://pub.dev/packages/schema_widget/score "Pub Points")
[![Popularity](https://badges.bar/schema_widget/popularity)](https://pub.dev/packages/schema_widget/score "Popularity")
[![style: effective dart](https://img.shields.io/badge/style-effective_dart-40c4ff.svg)](https://github.com/tenhobi/effective_dart "style: effective dart")
![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/Legytma/schema_widget "GitHub tag (latest SemVer)")
![GitHub top language](https://img.shields.io/github/languages/top/Legytma/schema_widget "GitHub top language")
![GitHub contributors](https://img.shields.io/github/contributors/Legytma/schema_widget "GitHub contributors")
![GitHub last commit](https://img.shields.io/github/last-commit/Legytma/schema_widget "GitHub last commit")
![Flutter CI](https://github.com/Legytma/schema_widget/workflows/Flutter%20CI/badge.svg "Flutter CI")

# Convert JSON to Widget validating with JSON Schema for Flutter apps

- [What is it](#what-is-it "What is it")
  - [Author](#author "Author")
  - [Motivation](#motivation "Motivation")
  - [How this work](#how-this-work "How this work")
    - [`TypeSchemaParser`](#typeschemaparser "TypeSchemaParser")
- [TYPE SCHEMA PARSER SUPPORT](TYPE_SCHEMA_PARSER_SUPPORT.md "TYPE SCHEMA PARSER SUPPORT")
- [Getting Started](#getting-started "Getting Started")
  - [Installation](#installation "Installation")
  - [Usage](#usage "Usage")
- [CHANGELOG](CHANGELOG.md "CHANGELOG")
- [Next steps](#next-steps "Next steps")
- [LICENSE](LICENSE "LICENSE")

## What is it

The [schema_widget](https://pub.dev/packages/schema_widget "schema_widget") is a Flutter package implemented with base on [dynamic_widget](https://pub.dev/packages/dynamic_widget "dynamic_widget") package that produces widgets dynamically interpreting JSON objects.

### Author

Writed by [Alex Manoel Ferreira Silva ![Support - Recommend/Endorse me on LinkedIn](https://img.shields.io/badge/Support-Recommed%2FEndorse%20me%20on%20Linkedin-blue?style=for-the-badge&logo=linkedin)](https://www.linkedin.com/in/windol/ "Support - Recommend/Endorse me on LinkedIn")

Under
[![Legytma Soluções Inteligentes](https://legytma.com.br/wp-content/uploads/2019/12/Legytma-Soluções-Inteligentes-e1599223582681.png) ![Company - Legytma Soluções Inteligentes in LinkedIn](https://img.shields.io/badge/Company-Legytma%20Solu%C3%A7%C3%B5es%20Inteligentes%20in%20LinkedIn-blue?style=for-the-badge&logo=linkedin)](https://www.linkedin.com/company/legytma "Company - Legytma Soluções Inteligentes in LinkedIn")

### Motivation

A major difficulty for any application developer is to ensure that all users keep their applications up to date to ensure the same user experience and to reduce the time required to fix bugs.

The most commonly used alternative to accelerate this process is [Code Push](https://github.com/Microsoft/code-push) which allows the application update without the need for a new deploy in the store. However in Code Push GitHub itself there is a [Code Push Support for Flutter](https://github.com/Microsoft/code-push/issues/624#issuecomment-532358395) request with comment saying that support depends on implementing dynamic update support in Flutter, there is also a reference to [Flutter Roadmap](https://github.com/flutter/flutter/wiki/Roadmap#changes) saying that support for This type of update is postponed according to the official comment [Code Push / Hot Update / out of band updates](https://github.com/flutter/flutter/issues/14330#issuecomment-485565194), which explains the reasons that led to the decision.

One possible solution to this deadlock is to implement a JSON interpreter that enables screen redesign, which can be obtained using [dynamic_widget](https://pub.dev/packages/dynamic_widget "dynamic_widget"). Although [dynamic_widget](https://pub.dev/packages/dynamic_widget "dynamic_widget") provides an easy way to extend support for other widgets through separate implementations. It limits the use of events and does not allow the use of complex business rules.

### How this work

The basic operation of [schema_widget](https://pub.dev/packages/schema_widget "schema_widget") is very similar to that of [dynamic_widget](https://pub.dev/packages/dynamic_widget "dynamic_widget"). Both interpret JSON and instantiate the widget for it.
The main difference with [schema_widget](https://pub.dev/packages/schema_widget "schema_widget") is that it allows the creation of parsers for any type of object and the use of pre-implemented rules. For this reason, it makes it possible to create much more complex business rules.

The [schema_widget](https://pub.dev/packages/schema_widget "schema_widget") package provides the `SchemaWidget` class, in most cases this will be the only class used. All methods in this class are static.

To add a new parser, the following methods can be used:

Default parsers: This method loads all the default parsers from [schema_widget](https://pub.dev/packages/schema_widget "schema_widget").

```dart
  SchemaWidget.registerParsers();
```

Synchronous: This method can be used to load a custom parser that can be loaded synchronously.

```dart
  TypeSchemaParser typeSchemaParserInstance = CustomTypeSchemaParser();

  SchemaWidget.registerTypeParser(typeSchemaParserInstance);
```

Asynchronous: This method can be used to load a custom parser that needs to be loaded asynchronously.

```dart
  String typeName = "Custom";
  Future<TypeSchemaParser> futureTypeSchemaParserInstance = () async => CustomTypeSchemaParser();

  SchemaWidget.registerTypeParserAsync(typeName, futureTypeSchemaParserInstance);
```

In addition to parsers, any type of function and object for use by parsers can also be registered. For this, the `SchemaWidget.registerLogic(logicName, logic)` method must be used. The logic name must be unique, otherwise the previously registered logic will be replaced.

Example:

```dart
  const string1 = "Hello World";
  const map1 = {"foo": "bar"};
  const object1 = Foo();
  var func1 = () => "Hello World";
  var asyncFunc1 = () async => "Hello World";
  var object2 = Bar();

  SchemaWidget.registerLogic("string1", string1);
  SchemaWidget.registerLogic("map1", map1);
  SchemaWidget.registerLogic("object1", object1);
  SchemaWidget.registerLogic("func1", func1);
  SchemaWidget.registerLogic("asyncFunc1", asyncFunc1);
  SchemaWidget.registerLogic("otherExemple", object2);
```

Both parsers and logics are registered using the [get_it](https://pub.dev/packages/get_it "get_it") package to manage everything.
After registering all the parsers and logic necessary for use. It may be necessary to wait for `GetIt` to finish loading asynchronous parsers. For this, `GetIt` makes `GetIt.I.allReady()` available for use in a `FutureBuilder` for the purpose of creating widgets only after all parsers are loaded as shown in the following example.

```dart
class MyApp extends StatelessWidget {
  MyApp({Key key}) : super(key: key) {
    SchemaWidget.registerParsers();
    SchemaWidget.registerLogic("funcHelloWorld", () => Text("Hello World"));
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: GetIt.I.allReady(ignorePendingAsyncCreation: false),
      builder: (BuildContext context, AsyncSnapshot snapshot) {
        if (!snapshot.hasData) {
          return CircularProgressIndicator();
        }

        return SchemaWidget.parse<Widget>(
          context,
          {
            "type": "Scaffold",
            "appBar": {
              "type": "AppBar",
              "title": {
                "type": "Text",
                "data": "Hello World Example",
              },
            },
            "body": {
              "type": "Center",
              "child": "funcHelloWorld",
            },
          },
        );
      },
    );
  }
}
```

Alternatively, `SchemaWidget` itself can be used to execute `SchemaWidget.registerParsers()` and `SchemaWidget.parse<Widget>()`, reducing the complexity of the code.

```dart
class MyApp extends StatelessWidget {
  MyApp({Key key}) : super(key: key) {
    SchemaWidget.registerLogic("funcHelloWorld", () => Text("Hello World"));
  }

  @override
  Widget build(BuildContext context) {
    return SchemaWidget(
      {
        "type": "Scaffold",
        "appBar": {
          "type": "AppBar",
          "title": {
            "type": "Text",
            "data": "Hello World Example",
          },
        },
        "body": {
          "type": "Center",
          "child": "funcHelloWorld",
        },
      },
    );
  }
}
```

#### `TypeSchemaParser`

Everything in [schema_widget](https://pub.dev/packages/schema_widget "schema_widget") is based on `TypeSchemaParser`.

`TypeSchemaParser` is the implementation of a JSON interpreter that converts it into a complex object.

## Getting Started

- For help over SchemaWidget usage, view the [example](https://github.com/Legytma/schema_form/tree/master/example);
- For help over class documentation, view the [documentation](https://pub.dev/documentation/schema_widget/latest);
- For help getting started with Flutter, view our online [documentation](https://flutter.io/);
- For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).

### Installation

- Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  get_it:
  schema_widget: ^2.0.0-beta.0
```

- Install packages from the command line:
  with Flutter:

```shell script
$ flutter packages get
```

- Import it into the code file:

```dart
 import 'package:schema_widget/schema_widget.dart';
```

### Usage

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:schema_widget/schema_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  final GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>();

  MyApp({Key key}) : super(key: key) {
    SchemaWidget.registerLogic(
      "onGenerateRoute",
      _onGenerateRoute,
    );
    SchemaWidget.registerLogic(
      "onUnknownRoute",
      _onUnknownRoute,
    );
    SchemaWidget.registerLogic(
      "navigatorKey",
      _navigatorKey,
    );
  }

  @override
  Widget build(BuildContext context) {
    return SchemaWidget(
      loadSchemas: false,
      {
        "type": "MaterialApp",
        "title": 'SchemaWidget Demo',
        "theme": {
          "primarySwatch": {
            "primary": 0xFF2196F3,
            "swatch": {
              "50": 0xFFE3F2FD,
              "100": 0xFFBBDEFB,
              "200": 0xFF90CAF9,
              "300": 0xFF64B5F6,
              "400": 0xFF42A5F5,
              "500": 0xFF2196F3,
              "600": 0xFF1E88E5,
              "700": 0xFF1976D2,
              "800": 0xFF1565C0,
              "900": 0xFF0D47A1,
            },
          },
        },
        "navigatorKey": "navigatorKey",
        "initialRoute": "home",
        "onGenerateRoute": "onGenerateRoute",
        "onUnknownRoute": "onUnknownRoute"
      },
    );
  }

  Route _onGenerateRoute(RouteSettings settings) {
    return MaterialPageRoute(
      builder: (buildContext) => MyHomePage(),
      settings: settings,
    );
  }

  Route _onUnknownRoute(RouteSettings settings) {
    return MaterialPageRoute(
      builder: (buildContext) => MyHomePage(),
      settings: settings.copyWith(name: "home"),
    );
  }
}

class MyHomePage extends StatelessWidget {
  MyHomePage({Key key}) : super(key: key) {
    SchemaWidget.registerLogic("funcHelloWorld", () => Text("Hello World"));
  }

  @override
  Widget build(BuildContext context) {
    return SchemaWidget(
      {
        "type": "Scaffold",
        "appBar": {
          "type": "AppBar",
          "title": {
            "type": "Text",
            "data": "Hello World Example",
          },
        },
        "body": {
          "type": "Center",
          "child": "funcHelloWorld",
        },
      },
    );
  }
}
```

## Next steps

- [x] Publish Package;
- [x] Make MVP;
- [x] Minimal documentation;
- [x] Change event binding of click event;
- [x] Add list of default supported Widgets;
- [x] Add list of default supported Types;
- [ ] Create content about;
- [ ] Publish complementar packages;
- [ ] Create example full functional apps;
- [ ] Make a commercial product using the package;


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0-beta.0](https://github.com/Legytma/schema_widget/compare/1.0.0...2.0.0-beta.0) (2020-09-11)


### Features

* support to channel beta on Flutter ([8f7cb30](https://github.com/Legytma/schema_widget/commit/8f7cb30767d3fe30cd4cdf99bb2b9c4e4fe38468))


### Changes that do not affect the meaning of the code

* reformat files ([fee4c16](https://github.com/Legytma/schema_widget/commit/fee4c1650f3d5dab3b99d3dd6c2c388da62e24b8))


### Code change that neither fixes a bug nor adds a feature

* update dependencies and schema version ([7908caf](https://github.com/Legytma/schema_widget/commit/7908caf1f9ef67983386bcc072620ff2df576672))

## [1.0.0](https://github.com/Legytma/schema_widget/compare/1.0.0-5...1.0.0) (2020-09-10)


### Documentation

* add badges of pub stats in README.md ([82fcd84](https://github.com/Legytma/schema_widget/commit/82fcd84691bca6e4fdcbdaa9309e311390f4dce0))
* import correction on README.md ([89f1866](https://github.com/Legytma/schema_widget/commit/89f18663bb808040afe606b66ec343e0a42efc13))


### Code change that neither fixes a bug nor adds a feature

* prepare project to stable release ([a0d59ab](https://github.com/Legytma/schema_widget/commit/a0d59ab94c552e99953a6b122c353d29815b9e40))
* update project example ([147c82e](https://github.com/Legytma/schema_widget/commit/147c82e50687351671708434b555b6fa55f2c745))
* update structure of example ([3b505ed](https://github.com/Legytma/schema_widget/commit/3b505edcd2ba6c32476569c59f7d437a0626bc5a))


### Changes that do not affect the meaning of the code

* change drawable logo ([7b32049](https://github.com/Legytma/schema_widget/commit/7b320497d5fd9ce63574facc7dc1b19a54806d7e))

## [1.0.0-5](https://github.com/Legytma/schema_widget/compare/1.0.0-4...1.0.0-5) (2020-08-07)


### Bug Fixes

* bug on generate imports ([5e50274](https://github.com/Legytma/schema_widget/commit/5e502744b6d81b71d0b0b43e238023ef92d081bc))


### Code change that neither fixes a bug nor adds a feature

* export json_schema_resolver.dart ([e2e2e9d](https://github.com/Legytma/schema_widget/commit/e2e2e9d38c32314aa23e8143bd864e269d993a26))
* replace LegytmaSchemas by schemas ([9ceaa97](https://github.com/Legytma/schema_widget/commit/9ceaa9713ea2278bdaf1edbb5643a57c78ae96bb))

## [1.0.0-4](https://github.com/Legytma/schema_widget/compare/1.0.0-3...1.0.0-4) (2020-08-05)


### Documentation

* fix TYPE_SCHEMA_PARSER_SUPPORT local URL ([6cf93a0](https://github.com/Legytma/schema_widget/commit/6cf93a0d1e9216d3f894386855f48370d26b9473))

## [1.0.0-3](https://github.com/Legytma/schema_widget/compare/1.0.0-2...1.0.0-3) (2020-08-05)


### Features

* add generator of supported parsers ([22bf3be](https://github.com/Legytma/schema_widget/commit/22bf3bec7cc1f1e0b11a679c0a730f03aebe76a0))


### Changes that do not affect the meaning of the code

* fix generators to beauty ([6ee9721](https://github.com/Legytma/schema_widget/commit/6ee97211000e3ae984e9169e18f695579f62338d))


### Documentation

* remove logs in example ([7a05d21](https://github.com/Legytma/schema_widget/commit/7a05d213715f820b633cac211e941ef3a172af38))

## [1.0.0-2](https://github.com/Legytma/schema_widget/compare/1.0.0-1...1.0.0-2) (2020-07-31)


### Documentation

* **license:** update license file ([be16a06](https://github.com/Legytma/schema_widget/commit/be16a067744bf7a04bf575afa914137c4f25114d))


### Code change that neither fixes a bug nor adds a feature

* adapt the project to the standards ([3039199](https://github.com/Legytma/schema_widget/commit/3039199805975122474149a57de51cc6b5784912))

## [1.0.0-1](https://github.com/Legytma/schema_widget/compare/1.0.0-0...1.0.0-1) (2020-07-30)


### Code change that neither fixes a bug nor adds a feature

* lock packages version ([586f50f](https://github.com/Legytma/schema_widget/commit/586f50f76c8cebea1930a1f4567fb45954229b69))

## [1.0.0-0](https://github.com/Legytma/schema_widget/compare/0.1.11...1.0.0-0) (2020-07-30)


### ⚠ BREAKING CHANGES

* Remove support to google_maps_flutter and migrate implementation to https://github.com/Legytma/google_maps_schema_parser.

Signed-off-by: Alex Manoel Ferreira Silva <alex@legytma.com.br>
* Remove support to flutter_spinkit and migrate implementation to https://github.com/Legytma/spinkit_schema_parser.

Signed-off-by: Alex Manoel Ferreira Silva <alex@legytma.com.br>

### Features

* add commit message hook ([4e41cdc](https://github.com/Legytma/schema_widget/commit/4e41cdc3f39c63fca250198576d828c8444ea293))
* add splash screen on example app ([3d0d35a](https://github.com/Legytma/schema_widget/commit/3d0d35acefc96e74341206a90eda9defc07da349))
* schemas version update to 2.0.0 ([173edd5](https://github.com/Legytma/schema_widget/commit/173edd50f55f400ce44e3738dbff537c5edb9261))


### Bug Fixes

* **lefthook:** pre-push command workaround ([9729183](https://github.com/Legytma/schema_widget/commit/9729183367f508f2049fdf8bb5bbc1057d6cfec4))


### Code changes that improves performance

* add eslint validation ([66f48d1](https://github.com/Legytma/schema_widget/commit/66f48d12bf9e9c00274c390f0c8a78dd6d910258))
* add lefthook support ([5fc367d](https://github.com/Legytma/schema_widget/commit/5fc367d9e3b24cd462f4f623adefd13830b9deb5))
* add pull request checks on github ([38bd1f3](https://github.com/Legytma/schema_widget/commit/38bd1f351f426318561fc22801786050db7183ff))
* adjust of commit analysis ([1fe0d38](https://github.com/Legytma/schema_widget/commit/1fe0d38ea18a31ce533f1475dffebc996ee84c2c))


### Documentation

* initial readme file update ([b2eebfb](https://github.com/Legytma/schema_widget/commit/b2eebfbc59457628528c45ec784e40901258274e))
* **license:** update license comments ([79eacac](https://github.com/Legytma/schema_widget/commit/79eacac1d22d4d3201d976638c2d3e80118c4215))

example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:logging/logging.dart';
import 'package:schema_widget/schema_widget.dart';
import 'package:schema_widget/widget/schema_widget_splash_screen.dart';

void main() {
  const _logIgnore = <String>[
    "JsonSchema",
    "JsonSchemaResolverStatistics",
    "FactoryStatistics",
//    "Validator",
  ];

  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen((rec) {
    if (_logIgnore.contains(rec.loggerName)) {
      return;
    }

    print('${rec.loggerName} - ${rec.level.name}: ${rec.time}: ${rec.message}');
    // developer.log(
    //   rec.message,
    //   level: rec.level.value,
    //   error: rec.error,
    //   name: rec.loggerName,
    //   sequenceNumber: rec.sequenceNumber,
    //   stackTrace: rec.stackTrace,
    //   time: rec.time,
    //   zone: rec.zone,
    // );
  });

  return runApp(MyApp());
}

/// App widget
class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final Logger _log = Logger("_MyAppState");

  final GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>();

  Route _onGenerateRoute(RouteSettings settings) {
    _log.finest("_onGenerateRoute -> settings: $settings");

    return MaterialPageRoute(
      builder: (buildContext) => MyHomePage(),
      settings: settings,
    );
  }

  Route _onUnknownRoute(RouteSettings settings) {
    _log.finest("_onUnknownRoute -> settings: $settings");

    return MaterialPageRoute(
      builder: (buildContext) => MyHomePage(),
      settings: settings.copyWith(name: "home"),
    );
  }

  @override
  void initState() {
    super.initState();

    SchemaWidget.registerLogic("onGenerateRoute", _onGenerateRoute);
    SchemaWidget.registerLogic("onUnknownRoute", _onUnknownRoute);
    SchemaWidget.registerLogic("navigatorKey", _navigatorKey);
  }

  @override
  Widget build(BuildContext context) {
    return SchemaWidget(
      {
        "type": "MaterialApp",
        "title": 'Flutter Demo',
        "theme": {
          "primarySwatch": {
            "primary": 0xFF2196F3,
            "swatch": {
              "50": 0xFFE3F2FD,
              "100": 0xFFBBDEFB,
              "200": 0xFF90CAF9,
              "300": 0xFF64B5F6,
              "400": 0xFF42A5F5,
              "500": 0xFF2196F3,
              "600": 0xFF1E88E5,
              "700": 0xFF1976D2,
              "800": 0xFF1565C0,
              "900": 0xFF0D47A1,
            },
          },
        },
        "navigatorKey": "navigatorKey",
        "initialRoute": "home",
        "onGenerateRoute": "onGenerateRoute",
        "onUnknownRoute": "onUnknownRoute"
      },
      loadParsers: true,
      loadSchemas: false,
      waitBuilder: (context) {
        _log.info("Loading SplashScreen...");
        return MaterialApp(
          home: Scaffold(
            body: SchemaWidgetSplashScreen(),
          ),
          theme: ThemeData(
            scaffoldBackgroundColor: Colors.white,
          ),
        );
      },
    );
  }
}

/// Home widget
class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  final StreamController<int> _streamControllerInt = StreamController<int>();

  _MyHomePageState() {
    SchemaWidget.registerLogic("streamInt", _streamControllerInt.stream);
    SchemaWidget.registerLogic("buildTextCounter", _buildTextCounter);
    SchemaWidget.registerLogic("incrementCounter", _incrementCounter);
  }

  void _incrementCounter() {
    _streamControllerInt.add(++_counter);
  }

  Widget _buildTextCounter(
      BuildContext buildContext, AsyncSnapshot<dynamic> snapshot) {
    var textStyle = Theme.of(buildContext).textTheme.headline4;

    return SchemaWidget({
      "type": "Text",
      "data": 'You have pushed the button this many times: ${snapshot.data}',
      "style": {
        "type": "TextStyle",
        "color": "#${textStyle.color.value.toRadixString(16)}",
        "debugLabel": textStyle.debugLabel,
        "decoration": textStyle.decoration.toString().substring(15),
        "fontFamily": textStyle.fontFamily,
        "fontSize": textStyle.fontSize,
        "fontWeight": textStyle.fontWeight.toString().substring(11),
        "fontStyle":
            textStyle.fontStyle == FontStyle.italic ? 'italic' : 'normal',
      },
    });
  }

  @override
  Widget build(BuildContext context) {
    return SchemaWidget({
      "type": "Scaffold",
      "appBar": {
        "type": "AppBar",
        "title": {
          "type": "Text",
          "data": "SchemaWidget Demo",
        },
      },
      "body": {
        "type": "Center",
        "child": {
          "type": "Column",
          "mainAxisAlignment": "center",
          "children": [
            {
              "type": "StreamBuilder<dynamic>",
              "stream": "streamInt",
              "initialData": 0,
              "builder": "buildTextCounter",
            },
          ],
        },
      },
      "floatingActionButton": {
        "type": "FloatingActionButton",
        "onPressed": "incrementCounter",
        "tooltip": 'Increment',
        "child": {
          "type": "Icon",
          "icon": {
            "codePoint": Icons.add.codePoint,
            "fontFamily": "MaterialIcons",
          },
        },
      },
    });
  }

  // @override
  // void dispose() {
  //   SchemaWidget.unregisterLogic("streamInt");
  //   SchemaWidget.unregisterLogic("buildTextCounter");
  //   SchemaWidget.unregisterLogic("incrementCounter");

  //   super.dispose();
  // }
}
